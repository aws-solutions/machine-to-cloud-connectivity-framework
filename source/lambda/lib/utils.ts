// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import axios from 'axios';
import { init } from '@paralleldrive/cuid2';
import { LambdaError } from './errors';
import { AnonymousMetricData, AwsSdkOptions } from './types/utils-types';

const METRICS_ENDPOINT = 'https://metrics.awssolutionsbuilder.com/generic';

/**
 * Generates an unique ID based on the parameter length.
 * @param length The length of the unique ID
 * @returns The unique ID
 */
export function generateUniqueId(length: number = 4) {
  const createId = init({
    length: length
  });
  return createId().toUpperCase();
}

/**
 * Gets the AWS JavaScript SDK options.
 * If the solution ID and version are provided, the options includes the custom user agent.
 * @param input The optional AWS SDK options
 * @returns The custom user agent options for AWS JavaScript SDK
 */
export function getAwsSdkOptions(input?: AwsSdkOptions): AwsSdkOptions {
  const { SOLUTION_ID, SOLUTION_VERSION } = process.env;
  const options: AwsSdkOptions = { ...input };

  if (SOLUTION_ID && SOLUTION_VERSION && SOLUTION_ID.trim() !== '' && SOLUTION_VERSION.trim() !== '') {
    options.customUserAgent = `AwsSolution/${SOLUTION_ID}/${SOLUTION_VERSION}`;
  }

  return options;
}

/**
 * Sends anonymous usage metrics.
 * @param data Data to send a anonymous metric
 * @param uuid The solution UUID
 */
export async function sendAnonymousMetric(data: AnonymousMetricData, uuid: string) {
  try {
    const { SOLUTION_ID, SOLUTION_VERSION } = process.env;
    const body = {
      Solution: SOLUTION_ID,
      Version: SOLUTION_VERSION,
      UUID: uuid,
      TimeStamp: new Date().toISOString().replace('T', ' ').replace('Z', ''),
      Data: data
    };

    const config = {
      headers: { 'Content-Type': 'application/json' }
    };

    await axios.post(METRICS_ENDPOINT, JSON.stringify(body), config);
  } catch (error) {
    console.error('Error sending an anonymous metric: ', error);
  }
}

/**
 * Sleeps for seconds.
 * @param seconds The seconds to sleep
 * @returns Promise to sleep for seconds
 */
export async function sleep(seconds: number): Promise<PromiseConstructor> {
  return new Promise(resolve => {
    setTimeout(resolve, seconds * 1000);
  });
}

/**
 * Trims all strings in the object or array.
 * @param obj The object to trim all strings
 * @returns The trimmed strings in the object
 * @throws `UtilsInvalidObject` when object is not provided
 */
export function trimAllStringInObjectOrArray(obj: unknown) {
  if (typeof obj !== 'object') {
    throw new LambdaError({
      message: 'Invalid object',
      name: 'UtilsInvalidObject',
      statusCode: 400
    });
  }

  let trimmedObject: unknown;
  if (Array.isArray(obj)) {
    trimmedObject = [];
  } else {
    trimmedObject = {};
  }

  for (const key in obj) {
    if (typeof obj[key] === 'object') {
      trimmedObject[key] = trimAllStringInObjectOrArray(obj[key]);
    } else {
      if (typeof obj[key] === 'string') {
        obj[key] = obj[key].trim();
      }

      if (Array.isArray(trimmedObject)) {
        trimmedObject.push(obj[key]);
      } else {
        trimmedObject[key] = obj[key];
      }
    }
  }

  return trimmedObject;
}

/**
 * Validates if the provided version is valid.
 * The valid version consists of a major version number, a minor version number, and a patch version number.
 * (e.g. 1.0.0, v1.0.0)
 * @param version The version to validate
 * @returns If the version is valid or not
 */
export function isValidVersion(version: string): boolean {
  return (
    /^(\d|[1-9]\d*)\.(\d|[1-9]\d*)\.(\d|[1-9]\d*)$/.test(version) ||
    /^v(\d|[1-9]\d*)\.(\d|[1-9]\d*)\.(\d|[1-9]\d*)$/.test(version)
  );
}

/**
 * Extracts the version from the version string.
 * The expected solution version is vX.Y.Z. X, Y, and Z are integer numbers.
 * When the solution version is v1.0.0, the component version would be 1.0.0.
 * If the solution version is not expected format (e.g. customVersion, v.1.0.0, v1, and so on),
 * the Lambda function is going to replace the component version to `1.0.0` by default.
 * @param version
 * @returns The version number
 */
export function extractValidVersion(version: string): string {
  if (isValidVersion(version)) {
    return version.replace(/^v/, '');
  } else {
    return '1.0.0';
  }
}
