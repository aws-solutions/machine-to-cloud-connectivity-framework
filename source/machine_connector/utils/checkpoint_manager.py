# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

from dbm import gnu as gdb
import json
import logging


class CheckpointManager:
    def __init__(self, checkpoint_db):
        self.checkpoint_db = checkpoint_db

        self.logger = logging.getLogger(self.__class__.__name__)
        self.logger.setLevel(logging.INFO)

    def read_checkpoint_db(self, stream_name: str) -> dict:
        try:
            with gdb.open(self.checkpoint_db, 'c') as db:
                self.checkpoint_entry = json.loads(db[stream_name])
            return self.checkpoint_entry
        except KeyError:
            self.logger.info(
                "Checkpoints for stream {} don't exist".format(stream_name))
            self.checkpoint_entry = {}
            return self.checkpoint_entry
        except Exception as err:
            self.logger.error(
                "Unable to access the checkpoint entry, {}".format(err))
            raise

    def write_checkpoint_db(self, stream_name: str, checkpoint_type: str, sequence: int) -> None:
        try:
            with gdb.open(self.checkpoint_db, 'c') as db:
                try:
                    self.checkpoint_entry = json.loads(db[stream_name])
                except KeyError:
                    self.checkpoint_entry = {}
                finally:
                    self.checkpoint_entry[checkpoint_type] = sequence
                    self.b_entry = json.dumps(
                        self.checkpoint_entry).encode('utf-8')
                    db[stream_name] = self.b_entry
        except Exception as err:
            self.logger.error("Unable to write {0} to {1}".format(
                err, self.checkpoint_db))
            raise

    def retrieve_checkpoints(self, stream_name: str):
        # Gets the latest checkpoints from the on-disk data store
        # If this is a new connection, there will be no checkpoints, return None
        try:
            self.checkpoints = self.read_checkpoint_db(stream_name)
            self.trailing_cp = self.checkpoints.get("trailing")
            self.primary_cp = self.checkpoints.get("primary")
            return self.trailing_cp, self.primary_cp
        except Exception as err:
            self.logger.error(
                "There was an issue retrieving checkpoints for stream {}: {}".format(stream_name, err))
            raise

    def write_checkpoints(self, stream_name: str, checkpoint: str, value: int) -> None:
        try:
            self.write_checkpoint_db(stream_name, checkpoint, value)
        except Exception as err:
            self.logger.error("There was an issue writing the checkpoint {} of value {} for stream {}: {}".format(
                checkpoint, value, stream_name, err))
            raise
