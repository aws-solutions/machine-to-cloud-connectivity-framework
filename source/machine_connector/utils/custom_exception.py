# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

class FileException(Exception):
    """Machine to Cloud Connectivity utils file exception"""
    pass


class ConnectorException(Exception):
    """Machine to Cloud Connectivity Framework connector exception"""
    pass


class ConverterException(Exception):
    """Machine to Cloud Connectivity Framework converter exception"""
    pass


class StreamManagerHelperException(Exception):
    """Machine to Cloud Connectivity Framework stream manager helper exception"""
    pass


class ValidationException(Exception):
    """Machine to Cloud Connectivity Framework validation exception"""
    pass


class PublisherException(Exception):
    """Machine to Cloud Connectivity Framework publisher exception"""
    pass
