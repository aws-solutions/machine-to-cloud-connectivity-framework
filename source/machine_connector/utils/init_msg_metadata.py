# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import os


class InitMessage:

    def __init__(self):
        # Site name from Greengrass Lambda Environment variables
        self.SITE_NAME = os.environ["SITE_NAME"]
        # Area from Greengrass Lambda Environment variables
        self.AREA = os.environ["AREA"]
        # Process from Greengrass Lambda Environment variables
        self.PROCESS = os.environ["PROCESS"]
        # Machine name from Greengrass Lambda Environment variables
        self.MACHINE_NAME = os.environ["MACHINE_NAME"]

    def init_user_message(self) -> dict:
        self.user_message = {}
        self.user_message["siteName"] = self.SITE_NAME
        self.user_message["area"] = self.AREA
        self.user_message["process"] = self.PROCESS
        self.user_message["machineName"] = self.MACHINE_NAME
        return (self.user_message)
