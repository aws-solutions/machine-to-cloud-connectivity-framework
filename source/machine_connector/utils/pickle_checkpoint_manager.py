# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import logging
import os
import pickle

"""
    This is a version of the checkpoint manager that is platform agnostic. It stores the sequence number of the
    respective Greengrass streams for pickup in case of restart or fall behind. Sequence numbers are
    stored like so:
    {
        "stream_name": {
            "checkpoint_type": sequence(1)
        }
    }

    Returns:
        _type_: _description_
"""


class PickleCheckpointManager:
    def __init__(self, checkpoint_file):
        self.checkpoint_file = checkpoint_file
        self._init_checkpoint_file()

        self.logger = logging.getLogger(self.__class__.__name__)
        self.logger.setLevel(logging.INFO)

    def _init_checkpoint_file(self):
        if not os.path.isfile(self.checkpoint_file):
            open(self.checkpoint_file, 'wb+').close()

    def _read_checkpoints(self) -> dict:
        all_checkpoints = {}
        if os.path.getsize(self.checkpoint_file) > 0:
            with open(self.checkpoint_file, 'rb') as file:
                all_checkpoints = pickle.load(file)  # nosec B301
        else:
            self.logger.warn('No checkpoints found during read')
        return all_checkpoints

    def read_checkpoint_db(self, stream_name: str) -> dict:
        try:
            all_checkpoints = self._read_checkpoints()
            return all_checkpoints[stream_name]
        except KeyError:
            self.logger.info(
                "Checkpoints for stream {} don't exist".format(stream_name))
            return {}
        except Exception as err:
            self.logger.error(
                "Unable to access the checkpoint entry, {}".format(err))
            raise

    def write_checkpoint_db(self, stream_name: str, checkpoint_type: str, sequence: int) -> None:
        try:
            all_checkpoints = self._read_checkpoints()
            if stream_name in all_checkpoints:
                if checkpoint_type in all_checkpoints[stream_name]:
                    all_checkpoints[stream_name][checkpoint_type] = sequence
                else:
                    all_checkpoints[stream_name] = {
                        checkpoint_type: sequence
                    }
            else:
                all_checkpoints[stream_name] = {
                    checkpoint_type: sequence
                }
            with open(self.checkpoint_file, 'wb') as write_file:
                pickle.dump(all_checkpoints, write_file)
        except Exception as err:
            self.logger.error("Unable to write {0} to {1}".format(
                err, self.checkpoint_file))
            raise

    def retrieve_checkpoints(self, stream_name: str):
        # Gets the latest checkpoints from the on-disk data store
        # If this is a new connection, there will be no checkpoints, return None
        try:
            checkpoints = self.read_checkpoint_db(stream_name)
            trailing_cp = checkpoints.get("trailing")
            primary_cp = checkpoints.get("primary")
            return trailing_cp, primary_cp
        except Exception as err:
            self.logger.error(
                "There was an issue retrieving checkpoints for stream {}: {}".format(stream_name, err))
            raise

    def write_checkpoints(self, stream_name: str, checkpoint: str, value: int) -> None:
        try:
            self.write_checkpoint_db(stream_name, checkpoint, value)
        except Exception as err:
            self.logger.error("There was an issue writing the checkpoint {} of value {} for stream {}: {}".format(
                checkpoint, value, stream_name, err))
            raise
