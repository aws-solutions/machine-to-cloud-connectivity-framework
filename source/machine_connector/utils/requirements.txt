backoff==2.1.2
awsiotsdk==1.11.1