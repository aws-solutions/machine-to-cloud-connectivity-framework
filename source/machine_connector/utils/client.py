# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

import logging
import json
import os
import awsiot.greengrasscoreipc

from awsiot.greengrasscoreipc.model import (
    PublishToIoTCoreRequest,
    QOS
)
from datetime import datetime
from utils.custom_exception import FileException
from utils.constants import WORK_BASE_DIR


class AWSEndpointClient:
    CONFIG_FILE_NAME = "connector-config.json"

    def __init__(self):
        # class general variables
        self.has_started = False
        self.is_running = False

        self.logger = logging.getLogger(self.__class__.__name__)
        self.logger.setLevel(logging.INFO)

        # Greengrass IPC client
        self.ipc_client = awsiot.greengrasscoreipc.connect()

    def start_client(self, connection_name: str, connection_configuration: dict) -> None:
        """
        Starts the connector client.
        """
        self.write_local_connection_configuration_file(
            connection_name=connection_name,
            connection_configuration=connection_configuration
        )
        self.has_started = True
        self.is_running = True

    def stop_client(self) -> None:
        """
        Stops the connector client.
        """
        self.is_running = False

    def publish_message_to_iot_topic(self, topic: str, payload: dict) -> None:
        """
        Publishes a message to the IoT topic.
        For more information, refer to
        https://docs.aws.amazon.com/greengrass/v2/developerguide/ipc-iot-core-mqtt.html#ipc-operation-publishtoiotcore

        :param topic: The IoT topic to publish the payload.
        :param payload: The payload to publish.
        """
        try:
            request = PublishToIoTCoreRequest()
            request.topic_name = topic
            request.payload = bytes(json.dumps(payload), "utf-8")
            request.qos = QOS.AT_MOST_ONCE

            operation = self.ipc_client.new_publish_to_iot_core()
            operation.activate(request)
        except Exception as err:
            self.logger.error(
                f"Failed to publish message to the IoT topic: {topic}. Error: {err}"
            )

    def read_local_connection_configuration(self, connection_name: str) -> dict:
        """
        Reads the local connection configuration file.

        :param connection_name: The connection name to get the local connection configuration.
        :return: The local configuration dictionary for the connection. If the file does not exist, return an empty dictionary.
        :raises: :err:`FileException` when any exception happens.
        """
        try:
            file_name = f"{WORK_BASE_DIR}/m2c2-{connection_name}/{self.CONFIG_FILE_NAME}"

            if os.path.exists(file_name):
                with open(file_name) as file:
                    return json.load(file)
            else:
                return {}
        except Exception as err:
            error_message = f"Failed to read the file: {err}"
            self.logger.error(error_message)
            raise FileException(error_message)

    def write_local_connection_configuration_file(self, connection_name: str, connection_configuration: dict) -> None:
        """
        Writes the local connection configuration file.

        :param connection_name: The connection name to write the local connection configuration.
        :param connection_configuration: The connection configuration to write locally.
        :raises: :err:`FileException` when any exception happens.
        """
        try:
            connection_configuration["_last-update-timestamp_"] = datetime.now().strftime(
                "%Y-%m-%d %H:%M:%S.%f")
            file_name = f"{WORK_BASE_DIR}/m2c2-{connection_name}/{self.CONFIG_FILE_NAME}"

            with open(file_name, "w+") as file:
                json.dump(connection_configuration, file, indent=2)
        except Exception as err:
            error_message = f"Failed to write to file: {err}"
            self.logger.error(error_message)
            raise FileException(error_message)
