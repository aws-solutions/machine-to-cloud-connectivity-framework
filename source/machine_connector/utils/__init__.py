# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

from .client import AWSEndpointClient
from .checkpoint_manager import CheckpointManager
from .stream_manager_helper import StreamManagerHelperClient
from .init_msg_metadata import InitMessage

__version__ = "4.1.0"
