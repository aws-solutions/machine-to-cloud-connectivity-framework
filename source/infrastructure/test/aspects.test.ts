// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Aspects, CfnCondition, CfnResource, Stack } from 'aws-cdk-lib';
import { Code, Function as LambdaFunction, Runtime } from 'aws-cdk-lib/aws-lambda';
import { ConditionAspect } from '../utils/aspects';

describe('ConditionAspect', () => {
  test('Add condition to all resources', () => {
    const stack = new Stack();
    const condition = new CfnCondition(stack, 'TestCondition');
    const lambdaFunction = new LambdaFunction(stack, 'TestLambdaFunction', {
      code: Code.fromInline(`console.log('Hello world!');`),
      handler: 'index.handler',
      runtime: Runtime.NODEJS_16_X
    });
    Aspects.of(stack).add(new ConditionAspect(condition));

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    const cfnResource = <CfnResource>lambdaFunction.node.defaultChild;
    expect(cfnResource.cfnOptions.condition).toEqual(condition);
  });
});
