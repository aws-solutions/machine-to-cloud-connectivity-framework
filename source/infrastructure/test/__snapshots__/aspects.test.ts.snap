// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionAspect Add condition to all resources 1`] = `
Object {
  "Resources": Object {
    "TestLambdaFunctionC089708A": Object {
      "Condition": "TestCondition",
      "DependsOn": Array [
        "TestLambdaFunctionServiceRole0C9E0634",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "console.log('Hello world!');",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestLambdaFunctionServiceRole0C9E0634",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestLambdaFunctionServiceRole0C9E0634": Object {
      "Condition": "TestCondition",
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
